### DISCUSSION TITLE
[FE302] 初探 useEffect ， localStorage 重新整理後資料消失問題

### DISCUSSION THREAD

**Discussion by rdcvgt** [Lidemy/forum#290] [前端相關問題] [answered]
參照影片中的寫法去執行，但是重新整理後，卻沒有顯示重新整理前已新增的 todo 
1. 新增 todo 時確定有寫入  localStorage
2. 重新整理時  `console.log("1 tododData", todoData)` 也還看得到已儲存的資料
3.  ` console.log("2 todos", todos)`  資料就消失了，只剩初始值
4.  懷疑是 setTodos 有出問題（沒更新 state），但我找不到目前程式碼有什麼狀況

附上照片（新增兩筆 todo 後重新整理，console 的狀態）
![截圖 2022-11-02 上午11 32 03](https://user-images.githubusercontent.com/75289113/199389010-836506b8-a449-469b-979c-d7a150f0fdbb.png)

這是我目前寫的程式碼
```javascript
function writeTodosToLocalStorage(todos) {
  if (todos.length !== 0) {
    window.localStorage.setItem("todos", JSON.stringify(todos))
    console.log("3 localStorage", window.localStorage)
  }
}


function App() {
  const [todos, setTodos] = useState([
    { id: 1, content: '上班', isDone: true },
    { id: 2, content: '睡覺', isDone: false }
  ])
  const [value, setValue] = useState("")
  const id = useRef(3)

  useEffect(() => {
    const todoData = (window.localStorage.getItem("todos")) || ""
    if (todoData) {
      console.log("1 tododData", todoData)
      setTodos(JSON.parse(todoData))
    }
  }, [])

  useEffect(() => {
    console.log("2 todos", todos)
    writeTodosToLocalStorage(todos)
  }, [todos])

//省略以下程式碼
}
```

**Comment by aszx87410** [comment, marked as answer]
可以用 https://codesandbox.io/ 弄一個簡單可重現的範例出來，會比較好 debug

**Reply by rdcvgt** [reply]
好的，再麻煩老師了！

https://codesandbox.io/s/interesting-panini-wt5eqr?file=/src/App.js

**Reply by aszx87410** [reply]
找到兇手囉，是 `<StrictMode>`，在這個模式底下有些 component 會被 render 兩遍，導致未預期的結果，拿掉就沒事了

參考資料：

1. https://stackoverflow.com/questions/60618844/react-hooks-useeffect-is-called-twice-even-if-an-empty-array-is-used-as-an-ar
2. https://stackoverflow.com/questions/61254372/my-react-component-is-rendering-twice-because-of-strict-mode

**Reply by rdcvgt** [reply]
謝謝老師！一直覺得一定是我漏寫什麼，結果沒想到是 `<StrictMode>` 的問題，非常感謝！！
